### ISSUE TITLE
Native Linux GUI support

### ISSUE THREAD

**Issue by Techwolf** [loot/loot#491]
When I tried to run LOOT after using the installer that had no errors, I get a blank window.
The console log:

```
steam@laptop ~ $ wine ~/.wine/drive_c/Program\ Files\ \(x86\)/LOOT/LOOT.exe
fixme:ole:RemUnknown_QueryInterface No interface for iid {00000019-0000-0000-c000-000000000046}
fixme:heap:RtlSetHeapInformation 0x110000 0 0x33fd30 4 stub
fixme:process:SetProcessDEPPolicy (3): stub
fixme:ver:GetCurrentPackageId (0x33f15c (nil)): stub
[2015-10-16 00:50:54.542631] [0x00000009] [trace]   Setting "TES IV: Oblivion" name to: TES IV: Oblivion
[2015-10-16 00:50:54.546992] [0x00000009] [trace]   Setting "TES IV: Oblivion" master file to: Oblivion.esm
[2015-10-16 00:50:54.547069] [0x00000009] [trace]   Setting "TES IV: Oblivion" repo URL to: https://github.com/loot/oblivion.git
[2015-10-16 00:50:54.547139] [0x00000009] [trace]   Setting "TES IV: Oblivion" repo branch to: v0.8
[2015-10-16 00:50:54.547213] [0x00000009] [trace]   Setting "TES IV: Oblivion" game path to: "~"
[2015-10-16 00:50:54.547285] [0x00000009] [trace]   Setting "TES IV: Oblivion" registry key to: Software\Bethesda Softworks\Oblivion\Installed Path
[2015-10-16 00:50:54.547367] [0x00000009] [trace]   Setting "TES V: Skyrim" name to: TES V: Skyrim
[2015-10-16 00:50:54.547439] [0x00000009] [trace]   Setting "TES V: Skyrim" master file to: Skyrim.esm
[2015-10-16 00:50:54.547509] [0x00000009] [trace]   Setting "TES V: Skyrim" repo URL to: https://github.com/loot/skyrim.git
[2015-10-16 00:50:54.547581] [0x00000009] [trace]   Setting "TES V: Skyrim" repo branch to: v0.8
[2015-10-16 00:50:54.547658] [0x00000009] [trace]   Setting "TES V: Skyrim" game path to: "C:\Program Files (x86)\Steam\steamapps\common\Skyrim\"
[2015-10-16 00:50:54.547729] [0x00000009] [trace]   Setting "TES V: Skyrim" registry key to: Software\Bethesda Softworks\Skyrim\Installed Path
[2015-10-16 00:50:54.547810] [0x00000009] [trace]   Setting "Fallout 3" name to: Fallout 3
[2015-10-16 00:50:54.547879] [0x00000009] [trace]   Setting "Fallout 3" master file to: Fallout3.esm
[2015-10-16 00:50:54.547949] [0x00000009] [trace]   Setting "Fallout 3" repo URL to: https://github.com/loot/fallout3.git
[2015-10-16 00:50:54.548020] [0x00000009] [trace]   Setting "Fallout 3" repo branch to: v0.8
[2015-10-16 00:50:54.548091] [0x00000009] [trace]   Setting "Fallout 3" game path to: "~"
[2015-10-16 00:50:54.548157] [0x00000009] [trace]   Setting "Fallout 3" registry key to: Software\Bethesda Softworks\Fallout3\Installed Path
[2015-10-16 00:50:54.548233] [0x00000009] [trace]   Setting "Fallout: New Vegas" name to: Fallout: New Vegas
[2015-10-16 00:50:54.548294] [0x00000009] [trace]   Setting "Fallout: New Vegas" master file to: FalloutNV.esm
[2015-10-16 00:50:54.548356] [0x00000009] [trace]   Setting "Fallout: New Vegas" repo URL to: https://github.com/loot/falloutnv.git
[2015-10-16 00:50:54.548418] [0x00000009] [trace]   Setting "Fallout: New Vegas" repo branch to: v0.8
[2015-10-16 00:50:54.548481] [0x00000009] [trace]   Setting "Fallout: New Vegas" game path to: "~"
[2015-10-16 00:50:54.548564] [0x00000009] [trace]   Setting "Fallout: New Vegas" registry key to: Software\Bethesda Softworks\FalloutNV\Installed Path
[2015-10-16 00:50:54.548638] [0x00000009] [trace]   Setting "TES IV: Oblivion" name to: Nehrim - At Fate's Edge
[2015-10-16 00:50:54.548699] [0x00000009] [trace]   Setting "Nehrim - At Fate's Edge" master file to: Nehrim.esm
[2015-10-16 00:50:54.548761] [0x00000009] [trace]   Setting "Nehrim - At Fate's Edge" repo URL to: https://github.com/loot/oblivion.git
[2015-10-16 00:50:54.548823] [0x00000009] [trace]   Setting "Nehrim - At Fate's Edge" repo branch to: v0.8
[2015-10-16 00:50:54.548888] [0x00000009] [trace]   Setting "Nehrim - At Fate's Edge" game path to: "~"
[2015-10-16 00:50:54.548950] [0x00000009] [trace]   Setting "Nehrim - At Fate's Edge" registry key to: Software\Microsoft\Windows\CurrentVersion\Uninstall\Nehrim - At Fate's Edge_is1\InstallLocation
fixme:module:load_library unsupported flag(s) used (flags: 0x00000800)
fixme:advapi:RegisterTraceGuidsW (0x624464, (nil), {f7b697a3-4db5-4d3b-be71-c4d284e6592f}, 7, 0x6e5900, (null), (null), 0x6f8c28): stub
fixme:advapi:RegisterTraceGuidsW   register trace class {72b14a7d-704c-423e-92f8-7e6d64bcb92a}
fixme:advapi:RegisterTraceGuidsW   register trace class {e2091f8a-1e0a-4731-84a2-0dd57c8a5261}
fixme:advapi:RegisterTraceGuidsW   register trace class {e8a3bf1f-a86b-4390-9c60-5390b969d22c}
fixme:advapi:RegisterTraceGuidsW   register trace class {5727a00f-50be-4519-8256-f7699871fecb}
fixme:advapi:RegisterTraceGuidsW   register trace class {7e854ec7-cdc4-405a-b5b2-aaf7c9e7d40c}
fixme:advapi:RegisterTraceGuidsW   register trace class {79a60dc6-5fc8-4952-a41c-1163aeec5eb8}
fixme:advapi:RegisterTraceGuidsW   register trace class {2718d25b-5bf5-4479-8e88-babc64bdbfca}
fixme:process:GetNumaHighestNodeNumber (0x33f5b4): semi-stub
fixme:heap:RtlSetHeapInformation (nil) 1 (nil) 0 stub
fixme:advapi:EventRegister {d2d578d9-2936-45b6-a09f-30e32715f42d}, 0x1018ddc0, 0x12e5b458, 0x12e8bec8
[1016/005054:WARNING:resource_bundle.cc(291)] locale_file_path.empty() for locale 
[1016/005054:ERROR:main_delegate.cc(688)] Could not load locale pak for en-US
[1016/005054:ERROR:main_delegate.cc(730)] Could not load cef_extensions.pak
fixme:imm:ImmDisableTextFrameService Stub
fixme:thread:GetThreadPreferredUILanguages 56, 0x33f904, (nil) 0x33f908
fixme:winsock:WSALookupServiceBeginW (0x33f680 0x00000ff0 0x33f6bc) Stub!
[1016/005054:ERROR:network_change_notifier_win.cc(153)] WSALookupServiceBegin failed with: 8
fixme:iphlpapi:NotifyAddrChange (Handle 0x33f820, overlapped 0x19f8b8): stub
fixme:win:RegisterDeviceNotificationW (hwnd=0x10054, filter=0x33f7f4,flags=0x00000000) returns a fake device notification handle!
fixme:win:RegisterDeviceNotificationW (hwnd=0x10054, filter=0x33f7f4,flags=0x00000000) returns a fake device notification handle!
fixme:win:EnumDisplayDevicesW ((null),0,0x33f270,0x00000000), stub!
[1016/005054:WARNING:resource_bundle.cc(489)] locale resources are not loaded
fixme:wtsapi:WTSRegisterSessionNotification Stub 0x10056 0x00000000
fixme:ole:RemUnknown_QueryInterface No interface for iid {00000019-0000-0000-c000-000000000046}
fixme:advapi:RegisterTraceGuidsW (0x10197940, 0x12e8bf08, {3dada31d-19ef-4dc1-b345-037927193422}, 1, 0x12e5b4ac, (null), (null), 0x12e8bf20): stub
fixme:advapi:RegisterTraceGuidsW   register trace class {00000000-0000-0000-0000-000000000000}
fixme:ntdll:NtQueryInformationToken QueryInformationToken( ..., TokenSessionId, ...) semi-stub
fixme:advapi:CreateRestrictedToken (0x258, 0x2, 7, 0x214b98, 20, 0x2127e8, 1, 0x1fdd38, 0x270da94): stub
fixme:advapi:CreateRestrictedToken (0x258, 0x2, 0, (nil), 0, (nil), 8, 0x213de0, 0x270da7c): stub
fixme:ntdll:NtSetInformationJobObject stub: 0x268 4 0x270e064 4
fixme:advapi:CreateProcessAsUserW 0x25c L"C:\\Program Files (x86)\\LOOT\\LOOT.exe" L"\"C:\\Program Files (x86)\\LOOT\\LOOT.exe\" --type=renderer --disable-databases --lang=en-US --lang=en-US --locales-dir-path=\"Z:\\home\\steam\\resources\\l10n\" --log-file=\"C:\\users\\steam\\Local Settings\\Application Data\\LOOT\\CEFDebugLog.txt\" --disable-spell-checking --device-scale-factor=1 -"... (nil) (nil) 0 0x0100040c (nil) (null) 0x270e1d0 0x270ddcc - semi-stub
fixme:module:load_library unsupported flag(s) used (flags: 0x00000800)
fixme:heap:RtlSetHeapInformation 0x240000 0 0x23fd90 4 stub
[1016/005054:ERROR:child_process_launcher.cc(444)] Failed to launch child process
```

I am using steam under wine to run skyrim. Steam has a lot of mods available for it. Would like to get LOOT working to mange the mod list.

**Comment by WrinklyNinja** [comment]
Looks like the Chromium part doesn't like running under Wine. It would be less effort to add native Linux support to the GUI, only the main application function, the clipboard interaction, and window size code need to have Linux-specific code added.

**Comment by WrinklyNinja** [comment]
@Techwolf: What Linux distribution/version are you using, and is it 32 or 64 bit?

**Comment by WrinklyNinja** [comment]
OK, I've managed to build the GUI and see it run (33ba0ebd0c61896cbeab8fa98e89ce8b9b973f56) on Ubuntu Desktop 14.04.3 64-bit.

I've still got script path assumptions to correct so that I can build the UI files and create archives. I've also got to make a note of all the external runtime dependencies.

The window title/size and clipboard stuff don't have any decent API available that I can work through, so they're too much effort to be worth fixing. I also won't be providing support for native Linux GUI builds, though once the scripts are fixed I may make snapshot builds of them available.

**Comment by WrinklyNinja** [comment]
@Techwolf: I've uploaded a [snapshot build](https://www.dropbox.com/sh/scuvwwc6ovzagmd/AAD1TodBAwGQTuV1-4Z2d0sCa?dl=0). It probably depends on the following packages: `libcurl4-openssl libx11 libboost-log1.55-dev libboost-date-time1.55 libboost-thread1.55 libboost-filesystem1.55 libboost-locale1.55 libboost-iostreams1.55 libssl`.

**Comment by Techwolf** [comment]
Thank you for the quick responce. Now that you said LOOT uses CEF, it may be this bug: https://bugs.winehq.org/show_bug.cgi?id=39403 that goes down to this re-opened bug: https://bugs.winehq.org/show_bug.cgi?id=21232

Not all is lost, just run CEF with --no-sandbox option.

I'll look at the snapshot later today. Only thing that may be a problem is different directorys due to wine and steam under wine usage. Meaning that even in windows, the files could be in two different location depending on how skyrim is installed, via DVD or steam.

**Comment by Techwolf** [comment]
Oh btw, I run Gentoo 64-bit and wine 32 and 64. I also maintain a gentoo overlay.

**Comment by Techwolf** [comment]
Just tried the snapshot build, get "./LOOT: error while loading shared libraries: libboost_log.so.1.55.0: cannot open shared object file: No such file or directory" May have to see about building it myself if the snapshop was based on current tip and not a separate branch.

**Comment by WrinklyNinja** [comment]
Yes, you need the shared Boost libraries, that's why I listed the packages above, but obviously on Gentoo things are different.

**Comment by WrinklyNinja** [comment]
To do:
- [ ] Document the build steps, which will mostly be "read the .travis.yml file", plus fetching and building CEF and the commands listed in 33ba0eb's message.
- [ ] File issues with a new `linux` label for the bits of functionality that are missing from Linux builds.

While I'd like to automate the builds like is done on Travis, the CEF binaries are locked behind CAPTCHAs. Because of this, I'll be leaving building binaries for various distributions to others.

I'll probably also leave fixing the bugs I file to others, unless there's some trivial addition I've missed.

**Comment by Freso** [comment]
I'm going to try and make a [PKGBUILD](https://wiki.archlinux.org/index.php/PKGBUILD) for the [AUR](https://wiki.archlinux.org/index.php/Arch_User_Repository). Should I make a separate ticket for any issues popping up, or post about it here?

**Comment by WrinklyNinja** [comment]
Just post about it here. If it turns out to be a big thing we can split it
off.

**Comment by Freso** [comment]
> clipboard stuff don't have any decent API available

There's [`xclip`](http://xclip.sourceforge.net/), which is a neat little command-line utility I often resort to in shell scripts etc. It can either be interfaced with directly, or you can poke through the code (it's GPL v2-or-later, so it can be copied willy-nilly into LOOT) and see how they deal with clipboards under X/Linux.

**Comment by Freso** [comment]
> Getting the game path from Registry (obviously). LOOT will need to be installed so that its folder sits beside the Data folder.

I get that it won't be possible to get game paths from Windows registry, but is there anything blocking setting the paths manually so you can have a single LOOT install instead of keeping one copy of LOOT per game?

**Comment by WrinklyNinja** [comment]
No, you can probably do that too in the GUI, it opens to the settings
window if it can't find any games.

**Comment by WrinklyNinja** [comment]
> There's xclip, which is a neat little command-line utility I often resort to in shell scripts etc. It can either be interfaced with directly, or you can poke through the code (it's GPL v2-or-later, so it can be copied willy-nilly into LOOT) and see how they deal with clipboards under X/Linux.

I'd rather call it than deal with X myself, but what about Wayland/Weston? From what I read it's only a year or so away from being production-ready.

**Comment by Freso** [comment]
One thing I just realised is: Does [CBash](https://github.com/lojack5/CBash) aim to work under Linux? While putting things together for Arch Linux, I noticed that libespm is scheduled to become deprecated once CBash is fully functional.

**Comment by WrinklyNinja** [comment]
Yeah... That'll never happen, I should probably update that message.

**Comment by Techwolf** [comment]
Just an update from me. I got started on some ebuilds for gentoo systems, right now, handling all the depens that loot requires. From an outsider view, it looks like the code from several projects should have been included in this project to simplify things.

I got an libespm ebuild done. It was nothing but headers anyway. Had to modified loot code to not use src/libespm.h, meaning /usr/include/libespm.h instead of /usr/include/src/libespm.h as "src" is not a good word to use in an include path, possible file collisions as "src" is a common word in linux. 

Libloadorder is done, but had to install some headers in /usr/include/api/*.h due to some loot code has #include <api/libloadorder.h>. Would have been better to use something like /usr/include/loadorder/*.h and till cmake to look in that directory for headers or fix the #include to point to it instead. Note that /usr/include can be different directories depending on the OS/system used.

As I said before, due to "src" and "api" used as include path, it appears these files was meant to be included in this project.

**Comment by Techwolf** [comment]
"Yes, you need the shared Boost libraries, that's why I listed the packages above, but obviously on Gentoo things are different."

The loot cmake code looks for static boost libs, but yet as you said above. Does it need both static and .so?

**Comment by Techwolf** [comment]
When I finally get loot built and running on my system, I am planning on creating a few cmake patches to improve things up. I have a bit of cmake experience due to working on Linden Labs secondlife viewer code and its cmake mess it has there. :-) ETA is when I have time.

**Comment by Techwolf** [comment]
"While I'd like to automate the builds like is done on Travis, the CEF binaries are locked behind CAPTCHAs. Because of this, I'll be leaving building binaries for various distributions to others."

I spent a couple hours, or so it seems like it, seeing if gentoo had a CEF ebuild already. I finally did find the CEF source. And cussed up a storm when the only compatible binary builds are behind some !@#$% javascripted site. My conclusion is the CEF depends is going to be the biggest PITA to get done.

Is it possible to really build loot without the CEF requirements? If so, is there command line docs about this?

**Comment by Freso** [comment]
LOOT's GUI *is* CEF, so I really highly doubt there's any way it's going to be possible to build without it. I just don't see how that would work. That'd be somewhat similar to building the GIMP without GTK+. (I also cursed at CEF when I tried to make a PKGBUILD for it. I decided to skip it for now and work on the other dependencies, and I will return to CEF once everything else is in place.)

**Comment by WrinklyNinja** [comment]
> From an outsider view, it looks like the code from several projects
should have been included in this project to simplify things.

> As I said before, due to "src" and "api" used as include path, it appears
these files was meant to be included in this project.

They originally were part of this code, but I try to split off the code
that's not LOOT-specific, it helps keep things clean and has modularity and
re-use advantages.

Your suggestions are correct though, the repository layouts are not
standard, but that's more a consequence of developing on Windows where
there aren't nice standard ways of doing things like that.

> The loot cmake code looks for static boost libs, but yet as you said
above. Does it need both static and .so?

No, I built with dynamic linking because in LOOT's configuration the choice
is tied to the choice of C++ runtime linking.

Linking is a bit of a tricky subject though, because ATM I'm already mixing
runtime versions on Linux, and that's not good.  On Windows everything is
built with MSVC 2013, but with Linux there's a mix of GCC 5 and 4.x, due to
the dependencies that I don't build myself.

> Is it possible to really build loot without the CEF requirements?

No, though it may be worth asking the Abobe Web Engine people or the CEF
maintainer if they can put at least some builds somewhere where downloads
can be scripted.

LOOT's GUI *is* CEF, so I really highly doubt there's any way it's going to
be possible to build without it. I just don't see how that would work.
That'd be somewhat similar to building the GIMP without GTK+. (I also
cursed at CEF when I tried to make a PKGBUILD for it. I decided to skip it
for now and work on the other dependencies, and I will return to CEF once
everything else is in place.)

—
Reply to this email directly or view it on GitHub
<https://github.com/loot/loot/issues/491#issuecomment-150903096>.

**Comment by WrinklyNinja** [comment]
Also, static linking on Linux would be nice to sort out, even if only so
that the masterlist CI wouldn't need to apt-get libraries to run the
validator. That's really a separate issue though...

**Comment by Freso** [comment]
Oh, btw @Techwolf,

> I got an libespm ebuild done. It was nothing but headers anyway. Had to modified loot code to not use src/libespm.h, meaning /usr/include/libespm.h instead of /usr/include/src/libespm.h as "src" is not a good word to use in an include path, possible file collisions as "src" is a common word in linux.

The [Arch Linux PKGBUILD](https://aur.archlinux.org/packages/libespm-git/) I made just copies the Git repository to /usr/include/libespm, so libespm.h ends up at /usr/include/libespm/src/libespm.h. Using the CMake flag `-DLIBESPM_ROOT=/usr/include/libespm` seems to deal with this properly. I'm open to doing things differently so Gentoo and Arch can be in sync on this. :)

> Libloadorder is done, but had to install some headers in /usr/include/api/*.h due to some loot code has `#include <api/libloadorder.h>`.

Again, my plan was to just place things in /usr/include/[lib]loadorder/…

Any chance you can link to your ebuilds for "inspiration"? :)

**Comment by WrinklyNinja** [comment]
I've updated the paths in libespm and libloadorder now, and updated LOOT's dev branch, though the Linux branch hasn't been rebased yet.

**Comment by Techwolf** [comment]
"Using the CMake flag -DLIBESPM_ROOT=/usr/include/libespm seems to deal with this properly." Ooo..I like this idea. Let me upload my prelim ebuilds for you somewhere.....

**Comment by Techwolf** [comment]
Tried to drag and drop like github says, but failes with "Unfortunately, we don’t support that file type.  Choose Files Try again with a PNG, GIF, JPG, DOCX, PPTX, XLSX, TXT, or PDF." So I'll just paste them here. They are by no mean close to done. I haven't even updated them to the new paths.

```
# Copyright 2015 Techwolf Lupindo
# Distributed under the terms of the GNU General Public License v2
# $Header: $

EAPI="5"

inherit git-r3

DESCRIPTION="A free software library for reading and writing .esp and .esm plugin files."
HOMEPAGE="https://github.com/WrinklyNinja/libespm"
EGIT_REPO_URI="https://github.com/WrinklyNinja/libespm.git"

LICENSE="GPL-3+"
SLOT="0"
KEYWORDS="~amd64 ~x86"
IUSE=""

src_install() {
	# Install headers
	cd "${S}/src"
	insinto /usr/include/
	doins *.h
	for i in fo3 fonv generic tes3 tes4 tes5; do
	  insinto /usr/include/$i
	  doins $i/*.h
	done
}
```

```
# Copyright 2015 Techwolf Lupindo
# Distributed under the terms of the GNU General Public License v2
# $Header: $

EAPI="5"

inherit git-r3 cmake-utils

DESCRIPTION="Libloadorder is a free software library for manipulating the load order and active status of plugins for TES III: Morrowind, TES IV: Oblivion, TES V: Skyrim, Fallout 3 and Fallout: New Vegas."
HOMEPAGE="https://github.com/WrinklyNinja/libloadorder"
EGIT_REPO_URI="https://github.com/WrinklyNinja/libloadorder.git"

LICENSE="GPL-3+"
SLOT="0"
KEYWORDS="~amd64 ~x86"
IUSE=""

DEPEND="dev-libs/boost[static-libs]
	dev-cpp/gtest
	dev-libs/libespm-git"

src_prepare() {
	# Don't prefix a bad path to includes.
	sed -i -e 's:#include <src/libespm.h>:#include <libespm.h>:' src/backend/plugins.h
	sed -i -e 's:#include <src/libespm.h>:#include <libespm.h>:' src/backend/game.h
	
	# skip the test for now, linking errors.
	sed -i -e 's:IF (${GTEST_FOUND}):IF (FALSE):' CMakeLists.txt
}

src_configure() {
	# export revision=$(get_version_component_range 4)
	mycmakeargs="${mycmakeargs} -DPROJECT_ARCH=64"
	cmake-utils_src_configure
}

src_install() {
	# Install header and libery
	cd ${S}/src
	insinto /usr/include/api
	doins api/*.h
	cd ${CMAKE_BUILD_DIR}
	dolib libloadorder64.a
}
```

```
# Copyright 2015 Techwolf Lupindo
# Distributed under the terms of the GNU General Public License v2
# $Header: $

EAPI="5"

inherit git-r3 cmake-utils

DESCRIPTION="LOOT is a plugin load order optimisation tool for TES IV: Oblivion, TES V: Skyrim, Fallout 3 and Fallout: New Vegas."
HOMEPAGE="https://github.com/loot/loot/tree/native-linux-gui"
EGIT_REPO_URI="https://github.com/loot/loot.git"
EGIT_BRANCH="native-linux-gui"

LICENSE="GPL-3+"
SLOT="0"
KEYWORDS="~amd64 ~x86"
IUSE=""

DEPEND="net-libs/nodejs
	dev-libs/boost[static-libs]
	dev-cpp/yaml-cpp
	dev-cpp/gtest"

src_configure() {
	# export revision=$(get_version_component_range 4)
	mycmakeargs="${mycmakeargs} -DPROJECT_ARCH=64"
	cmake-utils_src_configure
}
```

**Comment by Techwolf** [comment]
I just looked over the recent commits. Smart move baseing it off "/usr/include" sense it automatically included in the header search path of almost every build system out there. cmake automactelly add the system include path, no matter where it is. Not all distros use "/usr/include". No need to add cmake FindLIBespm.cmake and so on.

**Comment by WrinklyNinja** [comment]
I have now rebased the `native-linux-gui` branch off `dev`, and squashed some commits.

**Comment by Freso** [comment]
@Techwolf You will generally want to use [Gist](https://gist.github.com/) for uploading text files. It even supports syntax highlighting for .ebuilds: https://gist.github.com/Freso/69d3bcddbc0274b91e79 :)

**Comment by Techwolf** [comment]
@Freso I knew about gist, but didn't think of it at that time. My bad. :-)

**Comment by WrinklyNinja** [comment]
Note to self: do the following, then merge linux support into `dev`:

- [ ] Fix #494.
- [ ] Update README.md, BUILD.LINUX.md, and BUILD.MSVC.md so that there's no duplicated or missing information and everything's straightforward to follow.

**Comment by Techwolf** [comment]
Thank you for working on getting this ported/working on linux. This will give loot a leg up on the compition when SteamOS comes out this year. :-)

**Comment by Freso** [comment]
@Techwolf Except that SteamOS won't run TES3-5 or FO3/NV/4, so LOOT won't really be useful for SteamOS users… :) (It can potentially be useful for Lutris/POL/play-Steam-games-through-Wine users though, so it's not a completely futile experiment.)

**Comment by WrinklyNinja** [comment]
OK, closing this issue now that support has been merged into `dev`. If there's anything specific that doesn't already have an issue, open a new one for it.

**Comment by Techwolf** [comment]
Is there an updated linux build snapshot somewhere? The one snapshot posted here I managed to get running to the point of "(trusty64)steam@laptop:~/LOOT$ ./LOOT
./LOOT: error while loading shared libraries: libcef.so: cannot open shared object file: No such file or directory" and like to keep going in getting this program running on my system.

**Comment by WrinklyNinja** [comment]
> No, though it may be worth asking the Abobe Web Engine people or the CEF
> maintainer if they can put at least some builds somewhere where downloads
> can be scripted.

Following up on this, I received a reply to my query, and they've got no plans to support such downloads.

> Is there an updated linux build snapshot somewhere?

I have now uploaded a new [snapshot build](https://www.dropbox.com/sh/scuvwwc6ovzagmd/AAD1TodBAwGQTuV1-4Z2d0sCa?dl=0).

**Comment by Freso** [comment]
> they've got no plans to support such downloads.

Colour me not surprised…

**Comment by OUGHT** [comment]
Guys, what about Arch Linux PKGBUILD? It would be great if one could simply install LOOT from AUR instead of lurking about library replacements...

**Comment by WrinklyNinja** [comment]
I don't know about any distro-specific stuff, but building LOOT is now a lot easier than it was in December, as the process is almost all automated and self-contained - the only external dependency left is Boost, and that's kept external because it's widely-used and large enough that it's more efficient to use it that way.

**Comment by OUGHT** [comment]
Maybe. I'm trying it, but libloadorder does not want to build. =-)
```
[100%] Linking CXX executable tests
libloadorder.a(libloadorder.cpp.o): In function `lo_get_error_message':
libloadorder.cpp:(.text+0x10f): undefined reference to `TLS init function for liblo::extErrorString'
libloadorder.cpp:(.text+0x141): undefined reference to `TLS init function for liblo::extErrorString'
libloadorder.a(libloadorder.cpp.o): In function `lo_cleanup':
libloadorder.cpp:(.text+0x215): undefined reference to `TLS init function for liblo::extErrorString'
CMakeFiles/tests.dir/src/tests/main.cpp.o: In function `liblo::test::c_error_shouldSetTheGlobalErrorStringWhenPassedACodeAndString_Test::TestBody()':
main.cpp:(.text+0x13ec8): undefined reference to `TLS init function for liblo::extErrorString'
CMakeFiles/tests.dir/src/tests/main.cpp.o: In function `liblo::test::c_error_shouldSetTheGlobalErrorStringWhenPassedAnErrorObject_Test::TestBody()':
main.cpp:(.text+0x26921): undefined reference to `TLS init function for liblo::extErrorString'
```
The error stays even if I go into libloadorder folder and make actions according to travis.

**Comment by WrinklyNinja** [comment]
Are you using a C++11 compiler? I'm guessing TLS stands for thread-local storage, which is a C++11 feature. The Travis builds use GCC 5.4.

**Comment by OUGHT** [comment]
I am using standart Arch gcc version, which is GCC 6.1.1. Will try now to install GCC 5.4...

**Comment by WrinklyNinja** [comment]
I would have thought a later version would be fine too, but perhaps not.

**Comment by OUGHT** [comment]
The result is the same even at gcc-5:
```
[ 82%] Linking CXX executable tests
libloadorder.a(libloadorder.cpp.o): In function `lo_get_error_message':
libloadorder.cpp:(.text+0x10f): undefined reference to `TLS init function for liblo::extErrorString'
libloadorder.cpp:(.text+0x141): undefined reference to `TLS init function for liblo::extErrorString'
libloadorder.a(libloadorder.cpp.o): In function `lo_cleanup':
libloadorder.cpp:(.text+0x215): undefined reference to `TLS init function for liblo::extErrorString'
CMakeFiles/tests.dir/src/tests/main.cpp.o: In function `liblo::test::c_error_shouldSetTheGlobalErrorStringWhenPassedACodeAndString_Test::TestBody()':
main.cpp:(.text+0x13ec8): undefined reference to `TLS init function for liblo::extErrorString'
CMakeFiles/tests.dir/src/tests/main.cpp.o: In function `liblo::test::c_error_shouldSetTheGlobalErrorStringWhenPassedAnErrorObject_Test::TestBody()':
main.cpp:(.text+0x26921): undefined reference to `TLS init function for liblo::extErrorString'
```

**Comment by Freso** [comment]
I worked on an ArchLinux PKGBUILD back in the day (that is, about a year ago when this was being worked on initially; [libespm-git](https://aur.archlinux.org/packages/libespm-git/) and [lib32-boost-libs](https://aur.archlinux.org/packages/lib32-boost-libs/) are from that work), and what stomped me then was the CEF. Is the CEF any easier to deal with now?
